## **Preprocessing independent houses dataset**

In [33]:
# importing required dataset
import pandas as pd
import numpy as np
import re

In [34]:
data = pd.read_csv('../data/independent-house.csv')

In [35]:
data.head()

property_name  \
0    5 Bedroom House for sale in Sector 70A Gurgaon   
1  5 Bedroom House for sale in Sector 21A Faridabad   
2  10 Bedroom House for sale in Sushant Lok Phase 1   
3    21 Bedroom House for sale in Sector 54 Gurgaon   
4  12 Bedroom House for sale in Sushant Lok Phase 1   

                                                link           society  \
0  https://www.99acres.com/5-bhk-bedroom-independ...  Bptp Visionnaire   
1  https://www.99acres.com/5-bhk-bedroom-independ...               NaN   
2  https://www.99acres.com/10-bhk-bedroom-indepen...               NaN   
3  https://www.99acres.com/21-bhk-bedroom-indepen...               NaN   
4  https://www.99acres.com/12-bhk-bedroom-indepen...               NaN   

        price              rate                   area  \
0  5.25 Crore   ₹ 20,115/sq.ft.  (242 sq.m.) Plot Area   
1   5.7 Crore  ₹ 105,751/sq.ft.   (50 sq.m.) Plot Area   
2   2.1 Crore   ₹ 38,251/sq.ft.   (51 sq.m.) Plot Area   
3     5 Crore   ₹ 43,066/sq.ft.  (108 sq.m.) Plot Area   
4     3 Crore   ₹ 53,763/sq.ft.   (52 sq.m.) Plot Area   

                  areaWithType      bedRoom      bathroom       balcony  ...  \
0  Plot area 290(242.48 sq.m.)   5 Bedrooms   4 Bathrooms  3+ Balconies  ...   
1   Plot area 539(50.07 sq.m.)   5 Bedrooms   4 Bathrooms   2 Balconies  ...   
2       Plot area 61(51 sq.m.)  10 Bedrooms  10 Bathrooms  3+ Balconies  ...   
3  Plot area 129(107.86 sq.m.)  21 Bedrooms  21 Bathrooms  3+ Balconies  ...   
4    Plot area 62(51.84 sq.m.)  12 Bedrooms  12 Bathrooms  3+ Balconies  ...   

                                       address noOfFloor      facing  \
0    29b, Sector 70A Gurgaon, Gurgaon, Haryana  3 Floors  North-East   
1       Sector 21A Faridabad, Gurgaon, Haryana  2 Floors         NaN   
2        Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   
3          Sector 54 Gurgaon, Gurgaon, Haryana  5 Floors       North   
4  1228, Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   

      agePossession                                    nearbyLocations  \
0   0 to 1 Year Old  ['Good Earth City Center 2', 'Kunskapsskolan I...   
1  5 to 10 Year Old                                                NaN   
2   0 to 1 Year Old  ['Sector 42-43 metro station', 'Huda city cent...   
3   1 to 5 Year Old  ['Sector 53-54 metro station', 'Sector 54 chow...   
4   Within 6 months  ['Sector 42-43 metro station', 'Huda city cent...   

                                         description  \
0  Do you wish to buy an independent house in sec...   
1  Hi, we have an independent house/villa availab...   
2  Monthly rental income is rs1,40,000/- Best opt...   
3  129 sq yd plot size. 5 floors built up , fully...   
4  Best for investment purpose. Fully furnished b...   

                                      furnishDetails  \
0  ['1 Wardrobe', '1 Fan', '1 Exhaust Fan', '1 Ge...   
1  ['1 Water Purifier', '5 Fan', '1 Exhaust Fan',...   
2  ['10 Bed', '3 Fan', '10 Geyser', '2 Light', 'N...   
3  ['1 Water Purifier', '21 Fan', '1 Fridge', '1 ...   
4  ['1 Water Purifier', '1 Fridge', '12 Fan', '1 ...   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
1  ['Private Garden / Terrace', 'Park', 'Visitor ...   
2  ['Maintenance Staff', 'Water Storage', 'Visito...   
3  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
4  ['Maintenance Staff', 'Water Storage', 'Visito...   

                                              rating property_id  
0  ['Environment5 out of 5', 'Lifestyle4 out of 5...   K70037724  
1                                                NaN   E69288322  
2  ['Environment5 out of 5', 'Lifestyle5 out of 5...   F69536898  
3  ['Environment4 out of 5', 'Lifestyle5 out of 5...   R69483164  
4  ['Environment5 out of 5', 'Lifestyle5 out of 5...   M69381272  

[5 rows x 21 columns]

In [36]:
# checking the missing values
data.isnull().sum()

property_name        0
link                 0
society            624
price              127
rate                41
area                 0
areaWithType       108
bedRoom            108
bathroom           108
balcony            108
additionalRoom     506
address             64
noOfFloor          128
facing             421
agePossession      108
nearbyLocations    182
description         59
furnishDetails     352
features           421
rating             188
property_id         59
dtype: int64

In [37]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1095 entries, 0 to 1094
Data columns (total 21 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   property_name    1095 non-null   object
 1   link             1095 non-null   object
 2   society          471 non-null    object
 3   price            968 non-null    object
 4   rate             1054 non-null   object
 5   area             1095 non-null   object
 6   areaWithType     987 non-null    object
 7   bedRoom          987 non-null    object
 8   bathroom         987 non-null    object
 9   balcony          987 non-null    object
 10  additionalRoom   589 non-null    object
 11  address          1031 non-null   object
 12  noOfFloor        967 non-null    object
 13  facing           674 non-null    object
 14  agePossession    987 non-null    object
 15  nearbyLocations  913 non-null    object
 16  description      1036 non-null   object
 17  furnishDetails   743 non-null    

In [38]:
# checking for duplicates
data.duplicated().sum()

np.int64(13)

In [39]:
data.drop(columns=['link','property_id'], inplace=True)

In [40]:
data.head(2)

property_name           society  \
0    5 Bedroom House for sale in Sector 70A Gurgaon  Bptp Visionnaire   
1  5 Bedroom House for sale in Sector 21A Faridabad               NaN   

        price              rate                   area  \
0  5.25 Crore   ₹ 20,115/sq.ft.  (242 sq.m.) Plot Area   
1   5.7 Crore  ₹ 105,751/sq.ft.   (50 sq.m.) Plot Area   

                  areaWithType     bedRoom     bathroom       balcony  \
0  Plot area 290(242.48 sq.m.)  5 Bedrooms  4 Bathrooms  3+ Balconies   
1   Plot area 539(50.07 sq.m.)  5 Bedrooms  4 Bathrooms   2 Balconies   

                       additionalRoom  \
0                        Servant Room   
1  Store Room,Pooja Room,Servant Room   

                                     address noOfFloor      facing  \
0  29b, Sector 70A Gurgaon, Gurgaon, Haryana  3 Floors  North-East   
1     Sector 21A Faridabad, Gurgaon, Haryana  2 Floors         NaN   

      agePossession                                    nearbyLocations  \
0   0 to 1 Year Old  ['Good Earth City Center 2', 'Kunskapsskolan I...   
1  5 to 10 Year Old                                                NaN   

                                         description  \
0  Do you wish to buy an independent house in sec...   
1  Hi, we have an independent house/villa availab...   

                                      furnishDetails  \
0  ['1 Wardrobe', '1 Fan', '1 Exhaust Fan', '1 Ge...   
1  ['1 Water Purifier', '5 Fan', '1 Exhaust Fan',...   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
1  ['Private Garden / Terrace', 'Park', 'Visitor ...   

                                              rating  
0  ['Environment5 out of 5', 'Lifestyle4 out of 5...  
1                                                NaN

In [41]:
# checking the shape of the dataset
data.shape

(1095, 19)

## **Preprocessing**

In [42]:
data.head()

property_name           society  \
0    5 Bedroom House for sale in Sector 70A Gurgaon  Bptp Visionnaire   
1  5 Bedroom House for sale in Sector 21A Faridabad               NaN   
2  10 Bedroom House for sale in Sushant Lok Phase 1               NaN   
3    21 Bedroom House for sale in Sector 54 Gurgaon               NaN   
4  12 Bedroom House for sale in Sushant Lok Phase 1               NaN   

        price              rate                   area  \
0  5.25 Crore   ₹ 20,115/sq.ft.  (242 sq.m.) Plot Area   
1   5.7 Crore  ₹ 105,751/sq.ft.   (50 sq.m.) Plot Area   
2   2.1 Crore   ₹ 38,251/sq.ft.   (51 sq.m.) Plot Area   
3     5 Crore   ₹ 43,066/sq.ft.  (108 sq.m.) Plot Area   
4     3 Crore   ₹ 53,763/sq.ft.   (52 sq.m.) Plot Area   

                  areaWithType      bedRoom      bathroom       balcony  \
0  Plot area 290(242.48 sq.m.)   5 Bedrooms   4 Bathrooms  3+ Balconies   
1   Plot area 539(50.07 sq.m.)   5 Bedrooms   4 Bathrooms   2 Balconies   
2       Plot area 61(51 sq.m.)  10 Bedrooms  10 Bathrooms  3+ Balconies   
3  Plot area 129(107.86 sq.m.)  21 Bedrooms  21 Bathrooms  3+ Balconies   
4    Plot area 62(51.84 sq.m.)  12 Bedrooms  12 Bathrooms  3+ Balconies   

                       additionalRoom  \
0                        Servant Room   
1  Store Room,Pooja Room,Servant Room   
2                        Servant Room   
3                        Servant Room   
4                              Others   

                                       address noOfFloor      facing  \
0    29b, Sector 70A Gurgaon, Gurgaon, Haryana  3 Floors  North-East   
1       Sector 21A Faridabad, Gurgaon, Haryana  2 Floors         NaN   
2        Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   
3          Sector 54 Gurgaon, Gurgaon, Haryana  5 Floors       North   
4  1228, Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   

      agePossession                                    nearbyLocations  \
0   0 to 1 Year Old  ['Good Earth City Center 2', 'Kunskapsskolan I...   
1  5 to 10 Year Old                                                NaN   
2   0 to 1 Year Old  ['Sector 42-43 metro station', 'Huda city cent...   
3   1 to 5 Year Old  ['Sector 53-54 metro station', 'Sector 54 chow...   
4   Within 6 months  ['Sector 42-43 metro station', 'Huda city cent...   

                                         description  \
0  Do you wish to buy an independent house in sec...   
1  Hi, we have an independent house/villa availab...   
2  Monthly rental income is rs1,40,000/- Best opt...   
3  129 sq yd plot size. 5 floors built up , fully...   
4  Best for investment purpose. Fully furnished b...   

                                      furnishDetails  \
0  ['1 Wardrobe', '1 Fan', '1 Exhaust Fan', '1 Ge...   
1  ['1 Water Purifier', '5 Fan', '1 Exhaust Fan',...   
2  ['10 Bed', '3 Fan', '10 Geyser', '2 Light', 'N...   
3  ['1 Water Purifier', '21 Fan', '1 Fridge', '1 ...   
4  ['1 Water Purifier', '1 Fridge', '12 Fan', '1 ...   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
1  ['Private Garden / Terrace', 'Park', 'Visitor ...   
2  ['Maintenance Staff', 'Water Storage', 'Visito...   
3  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
4  ['Maintenance Staff', 'Water Storage', 'Visito...   

                                              rating  
0  ['Environment5 out of 5', 'Lifestyle4 out of 5...  
1                                                NaN  
2  ['Environment5 out of 5', 'Lifestyle5 out of 5...  
3  ['Environment4 out of 5', 'Lifestyle5 out of 5...  
4  ['Environment5 out of 5', 'Lifestyle5 out of 5...

In [43]:
# renaming the 'rate' column to 'price_per_sqft'
data.rename(columns={'rate':'price_per_sqft'}, inplace=True)

In [44]:
data['price_per_sqft'] = data['price_per_sqft'].str.split('/').str.get(0).str.replace('₹','').str.replace(',','').str.strip().astype('float')

In [45]:
data['price_per_sqft'].value_counts()

price_per_sqft
22222.0    14
11111.0    12
33333.0    11
10000.0    10
13333.0     7
           ..
13034.0     1
38150.0     1
28782.0     1
34491.0     1
10838.0     1
Name: count, Length: 776, dtype: int64

In [46]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1095 entries, 0 to 1094
Data columns (total 19 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   property_name    1095 non-null   object 
 1   society          471 non-null    object 
 2   price            968 non-null    object 
 3   price_per_sqft   1054 non-null   float64
 4   area             1095 non-null   object 
 5   areaWithType     987 non-null    object 
 6   bedRoom          987 non-null    object 
 7   bathroom         987 non-null    object 
 8   balcony          987 non-null    object 
 9   additionalRoom   589 non-null    object 
 10  address          1031 non-null   object 
 11  noOfFloor        967 non-null    object 
 12  facing           674 non-null    object 
 13  agePossession    987 non-null    object 
 14  nearbyLocations  913 non-null    object 
 15  description      1036 non-null   object 
 16  furnishDetails   743 non-null    object 
 17  features      

In [47]:
data['society'].value_counts()

society
Vipul Tatvam Villa                     27
Emaar MGF Marbella                     26
International City by SOBHA Phase 2    26
International City by Sobha Phase 1    23
Unitech Uniworld Resorts               14
                                       ..
Ansal Sushant Lok 2                     1
Sector 57 Huda                          1
Godrej Meridien                         1
Huda                                    1
Huda Floor                              1
Name: count, Length: 188, dtype: int64

In [48]:
data['society'] = data['society'].apply(lambda name: re.sub(r'\d+(\.\d+)?\s?★', '', str(name)).strip()).str.lower()

In [49]:
data[data['society']=='nan']

property_name society       price  \
1      5 Bedroom House for sale in Sector 21A Faridabad     nan   5.7 Crore   
2      10 Bedroom House for sale in Sushant Lok Phase 1     nan   2.1 Crore   
3        21 Bedroom House for sale in Sector 54 Gurgaon     nan     5 Crore   
4      12 Bedroom House for sale in Sushant Lok Phase 1     nan     3 Crore   
5      20 Bedroom House for sale in Sushant Lok Phase 1     nan   4.5 Crore   
...                                                 ...     ...         ...   
1082      5 Bedroom House for sale in Sector 46 Gurgaon     nan   3.6 Crore   
1083    5 Bedroom House for sale in Sushant Lok Phase 2     nan   3.1 Crore   
1084    3 Bedroom House for sale in Sushant Lok Phase 3     nan  4.75 Crore   
1092            8 Bedroom House for sale in DLF Phase 2     nan         NaN   
1093  8 Bedroom House for sale in C Block Sushant Lo...     nan         NaN   

      price_per_sqft                   area  \
1           105751.0   (50 sq.m.) Plot Area   
2            38251.0   (51 sq.m.) Plot Area   
3            43066.0  (108 sq.m.) Plot Area   
4            53763.0   (52 sq.m.) Plot Area   
5            39062.0  (107 sq.m.) Plot Area   
...              ...                    ...   
1082         24845.0  (135 sq.m.) Plot Area   
1083         20026.0  (144 sq.m.) Plot Area   
1084         28787.0  (149 sq.m.) Plot Area   
1092         23457.0  (376 sq.m.) Plot Area   
1093         31008.0  (180 sq.m.) Plot Area   

                                           areaWithType      bedRoom  \
1                            Plot area 539(50.07 sq.m.)   5 Bedrooms   
2                                Plot area 61(51 sq.m.)  10 Bedrooms   
3                           Plot area 129(107.86 sq.m.)  21 Bedrooms   
4                             Plot area 62(51.84 sq.m.)  12 Bedrooms   
5                           Plot area 128(107.02 sq.m.)  20 Bedrooms   
...                                                 ...          ...   
1082                        Plot area 161(134.62 sq.m.)   5 Bedrooms   
1083                        Plot area 172(143.81 sq.m.)   5 Bedrooms   
1084  Plot area 1600(148.64 sq.m.)Built Up area: 170...   3 Bedrooms   
1092                                                NaN          NaN   
1093                                                NaN          NaN   

          bathroom       balcony                      additionalRoom  \
1      4 Bathrooms   2 Balconies  Store Room,Pooja Room,Servant Room   
2     10 Bathrooms  3+ Balconies                        Servant Room   
3     21 Bathrooms  3+ Balconies                        Servant Room   
4     12 Bathrooms  3+ Balconies                              Others   
5     20 Bathrooms  3+ Balconies                        Servant Room   
...            ...           ...                                 ...   
1082   5 Bathrooms  3+ Balconies                        Servant Room   
1083   4 Bathrooms  3+ Balconies             Servant Room,Store Room   
1084   3 Bathrooms   3 Balconies                          Store Room   
1092           NaN           NaN                                 NaN   
1093           NaN           NaN                                 NaN   

                                                address noOfFloor      facing  \
1                Sector 21A Faridabad, Gurgaon, Haryana  2 Floors         NaN   
2                 Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   
3                   Sector 54 Gurgaon, Gurgaon, Haryana  5 Floors       North   
4           1228, Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   
5                 Sushant Lok Phase 1, Gurgaon, Haryana  4 Floors        East   
...                                                 ...       ...         ...   
1082                Sector 46 Gurgaon, Gurgaon, Haryana  3 Floors  South-East   
1083              Sushant Lok Phase 2, Gurgaon, Haryana  2 Floors  North-East   
1084  240 Sq.Yrd Old Siplex 3 Bhk Villa In Gated Blo...  2 Floors  No

In [50]:
data['society'] = data['society'].str.replace('nan','independent')

In [51]:
data['society'].value_counts()

society
independent                            624
vipul tatvam villa                      27
emaar mgf marbella                      26
international city by sobha phase 2     26
international city by sobha phase 1     23
                                      ... 
ansal sushant lok 2                      1
sector 57 huda                           1
godrej meridien                          1
huda                                     1
huda floor                               1
Name: count, Length: 186, dtype: int64

**`price` column**

In [52]:
data['price'].value_counts()

price
3.5 Crore     19
5.5 Crore     17
6.5 Crore     16
6 Crore       14
5 Crore       13
              ..
3.8 Crore      1
7.45 Crore     1
7.42 Crore     1
6.09 Crore     1
2.98 Crore     1
Name: count, Length: 302, dtype: int64

In [53]:
data = data[data['price']!='Price on Request']

In [54]:
# Creating a price convertor function
def price_converter(x):
    """
    This function converts price amounts into Crores.
    """
    # If the input is already a float, return it directly
    if isinstance(x, float):
        return x

    # Split the string into parts
    parts = x.split(' ')

    # Check if the split resulted in at least two parts
    if len(parts) < 2:
        return None  # Or handle it in another way, like returning 0

    # Extract the numeric value and unit
    numeric_value = float(parts[0])
    unit = parts[1]

    # Convert based on the unit
    if unit == 'Lac':
        return round(numeric_value / 100, 2)
    elif unit == 'Crore':
        return round(numeric_value, 2)
    else:
        return None  # Handle unexpected units

In [55]:
# Applying the price converter function to the price feature
data['price'] = data['price'].apply(price_converter)

In [56]:
data.head()

property_name           society  price  \
0    5 Bedroom House for sale in Sector 70A Gurgaon  bptp visionnaire   5.25   
1  5 Bedroom House for sale in Sector 21A Faridabad       independent   5.70   
2  10 Bedroom House for sale in Sushant Lok Phase 1       independent   2.10   
3    21 Bedroom House for sale in Sector 54 Gurgaon       independent   5.00   
4  12 Bedroom House for sale in Sushant Lok Phase 1       independent   3.00   

   price_per_sqft                   area                 areaWithType  \
0         20115.0  (242 sq.m.) Plot Area  Plot area 290(242.48 sq.m.)   
1        105751.0   (50 sq.m.) Plot Area   Plot area 539(50.07 sq.m.)   
2         38251.0   (51 sq.m.) Plot Area       Plot area 61(51 sq.m.)   
3         43066.0  (108 sq.m.) Plot Area  Plot area 129(107.86 sq.m.)   
4         53763.0   (52 sq.m.) Plot Area    Plot area 62(51.84 sq.m.)   

       bedRoom      bathroom       balcony  \
0   5 Bedrooms   4 Bathrooms  3+ Balconies   
1   5 Bedrooms   4 Bathrooms   2 Balconies   
2  10 Bedrooms  10 Bathrooms  3+ Balconies   
3  21 Bedrooms  21 Bathrooms  3+ Balconies   
4  12 Bedrooms  12 Bathrooms  3+ Balconies   

                       additionalRoom  \
0                        Servant Room   
1  Store Room,Pooja Room,Servant Room   
2                        Servant Room   
3                        Servant Room   
4                              Others   

                                       address noOfFloor      facing  \
0    29b, Sector 70A Gurgaon, Gurgaon, Haryana  3 Floors  North-East   
1       Sector 21A Faridabad, Gurgaon, Haryana  2 Floors         NaN   
2        Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   
3          Sector 54 Gurgaon, Gurgaon, Haryana  5 Floors       North   
4  1228, Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   

      agePossession                                    nearbyLocations  \
0   0 to 1 Year Old  ['Good Earth City Center 2', 'Kunskapsskolan I...   
1  5 to 10 Year Old                                                NaN   
2   0 to 1 Year Old  ['Sector 42-43 metro station', 'Huda city cent...   
3   1 to 5 Year Old  ['Sector 53-54 metro station', 'Sector 54 chow...   
4   Within 6 months  ['Sector 42-43 metro station', 'Huda city cent...   

                                         description  \
0  Do you wish to buy an independent house in sec...   
1  Hi, we have an independent house/villa availab...   
2  Monthly rental income is rs1,40,000/- Best opt...   
3  129 sq yd plot size. 5 floors built up , fully...   
4  Best for investment purpose. Fully furnished b...   

                                      furnishDetails  \
0  ['1 Wardrobe', '1 Fan', '1 Exhaust Fan', '1 Ge...   
1  ['1 Water Purifier', '5 Fan', '1 Exhaust Fan',...   
2  ['10 Bed', '3 Fan', '10 Geyser', '2 Light', 'N...   
3  ['1 Water Purifier', '21 Fan', '1 Fridge', '1 ...   
4  ['1 Water Purifier', '1 Fridge', '12 Fan', '1 ...   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
1  ['Private Garden / Terrace', 'Park', 'Visitor ...   
2  ['Maintenance Staff', 'Water Storage', 'Visito...   
3  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
4  ['Maintenance Staff', 'Water Storage', 'Visito...   

                                              rating  
0  ['Environment5 out of 5', 'Lifestyle4 out of 5...  
1                                                NaN  
2  ['Environment5 out of 5', 'Lifestyle5 out of 5...  
3  ['Environment4 out of 5', 'Lifestyle5 out of 5...  
4  ['Environment5 out of 5', 'Lifestyle5 out of 5...

**`area` column**

In [57]:
data['area'].value_counts()

area
(301 sq.m.) Plot Area    46
(84 sq.m.) Plot Area     40
(251 sq.m.) Plot Area    34
(167 sq.m.) Plot Area    27
(420 sq.m.) Plot Area    26
                         ..
(181 sq.m.) Plot Area     1
(509 sq.m.) Plot Area     1
(159 sq.m.) Plot Area     1
(234 sq.m.) Plot Area     1
(537 sq.m.) Plot Area     1
Name: count, Length: 384, dtype: int64

In [58]:
data['area'].str.split(' ').str.get(0).str.replace('(','').str.strip()

0       242
1        50
2        51
3       108
4        52
       ... 
1090    226
1091    226
1092    376
1093    180
1094     89
Name: area, Length: 1085, dtype: object

In [59]:
data.head()

property_name           society  price  \
0    5 Bedroom House for sale in Sector 70A Gurgaon  bptp visionnaire   5.25   
1  5 Bedroom House for sale in Sector 21A Faridabad       independent   5.70   
2  10 Bedroom House for sale in Sushant Lok Phase 1       independent   2.10   
3    21 Bedroom House for sale in Sector 54 Gurgaon       independent   5.00   
4  12 Bedroom House for sale in Sushant Lok Phase 1       independent   3.00   

   price_per_sqft                   area                 areaWithType  \
0         20115.0  (242 sq.m.) Plot Area  Plot area 290(242.48 sq.m.)   
1        105751.0   (50 sq.m.) Plot Area   Plot area 539(50.07 sq.m.)   
2         38251.0   (51 sq.m.) Plot Area       Plot area 61(51 sq.m.)   
3         43066.0  (108 sq.m.) Plot Area  Plot area 129(107.86 sq.m.)   
4         53763.0   (52 sq.m.) Plot Area    Plot area 62(51.84 sq.m.)   

       bedRoom      bathroom       balcony  \
0   5 Bedrooms   4 Bathrooms  3+ Balconies   
1   5 Bedrooms   4 Bathrooms   2 Balconies   
2  10 Bedrooms  10 Bathrooms  3+ Balconies   
3  21 Bedrooms  21 Bathrooms  3+ Balconies   
4  12 Bedrooms  12 Bathrooms  3+ Balconies   

                       additionalRoom  \
0                        Servant Room   
1  Store Room,Pooja Room,Servant Room   
2                        Servant Room   
3                        Servant Room   
4                              Others   

                                       address noOfFloor      facing  \
0    29b, Sector 70A Gurgaon, Gurgaon, Haryana  3 Floors  North-East   
1       Sector 21A Faridabad, Gurgaon, Haryana  2 Floors         NaN   
2        Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   
3          Sector 54 Gurgaon, Gurgaon, Haryana  5 Floors       North   
4  1228, Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   

      agePossession                                    nearbyLocations  \
0   0 to 1 Year Old  ['Good Earth City Center 2', 'Kunskapsskolan I...   
1  5 to 10 Year Old                                                NaN   
2   0 to 1 Year Old  ['Sector 42-43 metro station', 'Huda city cent...   
3   1 to 5 Year Old  ['Sector 53-54 metro station', 'Sector 54 chow...   
4   Within 6 months  ['Sector 42-43 metro station', 'Huda city cent...   

                                         description  \
0  Do you wish to buy an independent house in sec...   
1  Hi, we have an independent house/villa availab...   
2  Monthly rental income is rs1,40,000/- Best opt...   
3  129 sq yd plot size. 5 floors built up , fully...   
4  Best for investment purpose. Fully furnished b...   

                                      furnishDetails  \
0  ['1 Wardrobe', '1 Fan', '1 Exhaust Fan', '1 Ge...   
1  ['1 Water Purifier', '5 Fan', '1 Exhaust Fan',...   
2  ['10 Bed', '3 Fan', '10 Geyser', '2 Light', 'N...   
3  ['1 Water Purifier', '21 Fan', '1 Fridge', '1 ...   
4  ['1 Water Purifier', '1 Fridge', '12 Fan', '1 ...   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
1  ['Private Garden / Terrace', 'Park', 'Visitor ...   
2  ['Maintenance Staff', 'Water Storage', 'Visito...   
3  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
4  ['Maintenance Staff', 'Water Storage', 'Visito...   

                                              rating  
0  ['Environment5 out of 5', 'Lifestyle4 out of 5...  
1                                                NaN  
2  ['Environment5 out of 5', 'Lifestyle5 out of 5...  
3  ['Environment4 out of 5', 'Lifestyle5 out of 5...  
4  ['Environment5 out of 5', 'Lifestyle5 out of 5...

In [60]:
data['area']

0       (242 sq.m.) Plot Area
1        (50 sq.m.) Plot Area
2        (51 sq.m.) Plot Area
3       (108 sq.m.) Plot Area
4        (52 sq.m.) Plot Area
                ...          
1090    (226 sq.m.) Plot Area
1091    (226 sq.m.) Plot Area
1092    (376 sq.m.) Plot Area
1093    (180 sq.m.) Plot Area
1094     (89 sq.m.) Plot Area
Name: area, Length: 1085, dtype: object

#### `bedroom` column

In [61]:
data['bedRoom'].value_counts()

bedRoom
4 Bedrooms     239
5 Bedrooms     195
3 Bedrooms     148
2 Bedrooms      88
6 Bedrooms      82
9 Bedrooms      45
1 Bedroom       38
8 Bedrooms      34
7 Bedrooms      31
12 Bedrooms     29
10 Bedrooms     21
16 Bedrooms     12
13 Bedrooms      4
11 Bedrooms      2
19 Bedrooms      2
18 Bedrooms      2
21 Bedrooms      1
20 Bedrooms      1
36 Bedrooms      1
34 Bedrooms      1
14 Bedrooms      1
Name: count, dtype: int64

In [62]:
data[data['bedRoom'].isna()]

property_name  \
60        4 Bedroom House for sale in Sector 86 Gurgaon   
61    8 Bedroom House for sale in B1 Block Sector 57...   
62              6 Bedroom House for sale in DLF Phase 1   
63      6 Bedroom House for sale in Sushant Lok Phase 3   
64      6 Bedroom House for sale in Sushant Lok Phase 3   
...                                                 ...   
894       6 Bedroom House for sale in Sector 22 Gurgaon   
1091                 3 Bedroom House for sale in Bissar   
1092            8 Bedroom House for sale in DLF Phase 2   
1093  8 Bedroom House for sale in C Block Sushant Lo...   
1094      2 Bedroom House for sale in Sector 92 Gurgaon   

                                       society  price  price_per_sqft  \
60                        dlf new town heights    NaN          7718.0   
61                                 independent    NaN         25535.0   
62                                 independent    NaN          1859.0   
63                     ansal sushant residency    NaN         17037.0   
64                       ansals florence villa    NaN         10579.0   
...                                        ...    ...             ...   
894                saksham welfare association    NaN         19645.0   
1091  silverglades tarudhan valley golf resort    NaN         13992.0   
1092                               independent    NaN         23457.0   
1093                               independent    NaN         31008.0   
1094                  signature global city 92    NaN         13034.0   

                       area areaWithType bedRoom bathroom balcony  \
60    (297 sq.m.) Plot Area          NaN     NaN      NaN     NaN   
61    (135 sq.m.) Plot Area          NaN     NaN      NaN     NaN   
62    (420 sq.m.) Plot Area          NaN     NaN      NaN     NaN   
63    (251 sq.m.) Plot Area          NaN     NaN      NaN     NaN   
64    (321 sq.m.) Plot Area          NaN     NaN      NaN     NaN   
...                     ...          ...     ...      ...     ...   
894   (220 sq.m.) Plot Area          NaN     NaN      NaN     NaN   
1091  (226 sq.m.) Plot Area          NaN     NaN      NaN     NaN   
1092  (376 sq.m.) Plot Area          NaN     NaN      NaN     NaN   
1093  (180 sq.m.) Plot Area          NaN     NaN      NaN     NaN   
1094   (89 sq.m.) Plot Area          NaN     NaN      NaN     NaN   

     additionalRoom address noOfFloor facing agePossession nearbyLocations  \
60              NaN     NaN       NaN    NaN           NaN             NaN   
61              NaN     NaN       NaN    NaN           NaN             NaN   
62              NaN     NaN       NaN    NaN           NaN             NaN   
63              NaN     NaN       NaN    NaN           NaN             NaN   
64              NaN     NaN       NaN    NaN           NaN             NaN   
...             ...     ...       ...    ...           ...             ...   
894             NaN     NaN       NaN    NaN           NaN             NaN   
1091            NaN     NaN       NaN    NaN           NaN             NaN   
1092            NaN     NaN       NaN    NaN           NaN             NaN   
1093            NaN     NaN       NaN    NaN           NaN             NaN   
1094            NaN     NaN       NaN    NaN           NaN             NaN   

     description furnishDetails features rating  
60           NaN            NaN      NaN    NaN  
61           NaN            NaN      NaN    NaN  
62           NaN            NaN      NaN    NaN  
63           NaN            NaN      NaN    NaN  
64           NaN            NaN      NaN    NaN  
...          ...            ...      ...    ...  
894          NaN            NaN      NaN    NaN  
1091         NaN            NaN      NaN    NaN  
1092         NaN            NaN      NaN    NaN  
1093         NaN            NaN      NaN    NaN  
1094         NaN            NaN      NaN    NaN  

[108 rows x 19 columns]

There are 108 rows that are having 78.9% of missing values. Thus removing these rows.

In [63]:
data = data[data['bedRoom'].notnull()]

In [64]:
data[data['bedRoom'].notnull()]

property_name  \
0       5 Bedroom House for sale in Sector 70A Gurgaon   
1     5 Bedroom House for sale in Sector 21A Faridabad   
2     10 Bedroom House for sale in Sushant Lok Phase 1   
3       21 Bedroom House for sale in Sector 54 Gurgaon   
4     12 Bedroom House for sale in Sushant Lok Phase 1   
...                                                ...   
1086     4 Bedroom House for sale in Sector 26 Gurgaon   
1087           3 Bedroom House for sale in DLF Phase 2   
1088     3 Bedroom House for sale in Sector 26 Gurgaon   
1089     3 Bedroom House for sale in Sector 26 Gurgaon   
1090                3 Bedroom House for sale in Bissar   

                                       society  price  price_per_sqft  \
0                             bptp visionnaire   5.25         20115.0   
1                                  independent   5.70        105751.0   
2                                  independent   2.10         38251.0   
3                                  independent   5.00         43066.0   
4                                  independent   3.00         53763.0   
...                                        ...    ...             ...   
1086                          dlf city phase 1   5.50         30556.0   
1087                    dlf city plots phase 2   4.25         31481.0   
1088                          dlf city phase 1   4.50         33333.0   
1089                          dlf city phase 1   3.25         33129.0   
1090  silverglades tarudhan valley golf resort   3.60         14815.0   

                       area                 areaWithType      bedRoom  \
0     (242 sq.m.) Plot Area  Plot area 290(242.48 sq.m.)   5 Bedrooms   
1      (50 sq.m.) Plot Area   Plot area 539(50.07 sq.m.)   5 Bedrooms   
2      (51 sq.m.) Plot Area       Plot area 61(51 sq.m.)  10 Bedrooms   
3     (108 sq.m.) Plot Area  Plot area 129(107.86 sq.m.)  21 Bedrooms   
4      (52 sq.m.) Plot Area    Plot area 62(51.84 sq.m.)  12 Bedrooms   
...                     ...                          ...          ...   
1086  (167 sq.m.) Plot Area  Plot area 200(167.23 sq.m.)   4 Bedrooms   
1087  (125 sq.m.) Plot Area  Plot area 150(125.42 sq.m.)   3 Bedrooms   
1088  (125 sq.m.) Plot Area  Plot area 150(125.42 sq.m.)   3 Bedrooms   
1089   (91 sq.m.) Plot Area   Plot area 109(91.14 sq.m.)   3 Bedrooms   
1090  (226 sq.m.) Plot Area  Plot area 270(225.75 sq.m.)   3 Bedrooms   

          bathroom       balcony                             additionalRoom  \
0      4 Bathrooms  3+ Balconies                               Servant Room   
1      4 Bathrooms   2 Balconies         Store Room,Pooja Room,Servant Room   
2     10 Bathrooms  3+ Balconies                               Servant Room   
3     21 Bathrooms  3+ Balconies                               Servant Room   
4     12 Bathrooms  3+ Balconies                                     Others   
...            ...           ...                                        ...   
1086   4 Bathrooms   3 Balconies         Pooja Room,Study Room,Servant Room   
1087   2 Bathrooms   3 Balconies                                 Study Room   
1088   3 Bathrooms   2 Balconies                    Study Room,Servant Room   
1089   3 Bathrooms   3 Balconies                                 Study Room   
1090   3 Bathrooms   2 Balconies  Pooja Room,Study Room,Servant Room,Others   

                                          address noOfFloor      facing  \
0       29b, Sector 70A Gurgaon, Gurgaon, Haryana  3 Floors  North-East   
1          Sector 21A Faridabad, Gurgaon, Haryana  2 Floors         NaN   
2           Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   
3             Sector 54 Gurgaon, Gurgaon, Haryana  5 Floors       North   
4     1228, Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   
...                                           ...       ...         ...   
1086          Sector 26 Gurgaon, Gurgaon, Haryana  2 Floors  North-East   
1087                DLF Phase 2, Gurgaon,